### ISSUE TITLE
invalid-assignment false positive when assigning to attributes of Mapped[T | None] after None-check and local variable assignment

### ISSUE THREAD

**Issue by N1ntek** [astral-sh/ty#3004]
### Summary

When using SQLAlchemy's Mapped[T | None] for a relationship attribute, ty emits invalid-assignment errors even after:

A None check with assignment of a new instance
Storing the attribute in a local variable 

## MRE
```py
from sqlalchemy.orm import DeclarativeBase, Mapped, relationship, mapped_column


class Base(DeclarativeBase):
    pass


class Inner(Base):
    __tablename__ = "inner"
    id: Mapped[int] = mapped_column(primary_key=True)
    value: Mapped[int] = mapped_column(default=0)


class Outer(Base):
    __tablename__ = "outer"
    id: Mapped[int] = mapped_column(primary_key=True)
    inner: Mapped[Inner | None] = relationship("Inner")


def update(obj: Outer) -> None:
    if obj.inner is None:
        obj.inner = Inner()

    local = obj.inner  # expected: Inner, actual: Inner | None
    local.value = 42  # error[invalid-assignment]
```

## Expected behavior

After assigning `Inner()` to `obj.inner` inside the `if` block, and then storing it in a local variable `local`, ty should narrow `local` to `Inner` (not `Inner | None`).

## Actual behavior
error[invalid-assignment]: Object of type `int` is not assignable to attribute `value` on type `Inner | None`

### Version

ty 0.0.21 (c1ad9f281 2026-03-06)

**Comment by sharkdp** [comment]
Thank you for reporting this.

I assume this is related to the fact that `Mapped` is a descriptor. We normally support narrowing by attribute assignment.

**Comment by carljm** [comment]
Yeah, we are more conservative about descriptor narrowing than other type checkers currently. This is tracked in #628.
